src/utils/offlineRequestHandler.ts

```diff
@@ -0,0 +1,268 @@
+    } else {
+      // Assume it's text or JSON
+      return { type: 'text', body: await body.text() };
+    }
+  }
+
+  // Replay requests from IndexedDB
+  public async replayRequests() {
+    const db = await this.openDatabase();
+    const tx = db.transaction(this.storeName, 'readwrite');
+    const store = tx.objectStore(this.storeName);
+
+    const requests: SerializedRequest[] = [];
+
+    await new Promise<void>((resolve) => {
+      store.openCursor().onsuccess = (event: Event) => {
+        const cursor = (event.target as IDBRequest<IDBCursorWithValue | null>)
+          .result;
+        if (cursor) {
+          requests.push(cursor.value as SerializedRequest);
+          cursor.continue();
+        } else {
+          // All entries have been collected, resolve the promise
+          resolve();
+        }
+      };
+    });
+
+    requests.sort((a, b) => a.timestamp - b.timestamp);
+
+    for (const request of requests) {
+      const { url, method, headers, body, bodyType } = request;
+      const reconstructedBody = this.reconstructBody(body, bodyType);
+      const fetchOptions: RequestInit = {
+        method,
+        headers: JSON.parse(headers),
+        body: reconstructedBody
+      };
+
+      let attempts = 0;
+      let success = false;
+
+      while (!success && attempts < this.maxRetryAttempts) {
+        try {
+          const response = await fetch(url, fetchOptions);
+          if (response.ok) {
+            success = true;
+          } else {
+            attempts++;
+            await this.delay(this.retryDelayMs);
+          }
+        } catch (error) {
+          console.error(`Failed to replay request: ${url}`, error);
+          attempts++;
+          await this.delay(this.retryDelayMs);
+        }
+
+        if (attempts === this.maxRetryAttempts) {
+          console.log(`Max retry attempts reached for request url: ${url}`);
+          break;
+        }
+      }
+    }
+
+    try {
+      await db
+        .transaction(this.storeName, 'readwrite')
+        .objectStore(this.storeName)
+        .clear();
+    } catch (error) {
+      console.error('Error clearing IndexedDB:', error);
+    }
+  }
+
+  private delay(ms: number): Promise<void> {
+    return new Promise((resolve) => setTimeout(resolve, ms));
+  }
+
+  private reconstructBody(
+    body: any,
+    bodyType: 'blob' | 'formData' | 'arrayBuffer' | 'text'
+  ): Blob | FormData | null {
+    switch (bodyType) {
+      case 'blob':
+        return new Blob([body]);
+      case 'formData': {
+        const formData = new FormData();
+        const parsedBody = JSON.parse(body) as Record<string, any>;
+        Object.entries(parsedBody).forEach(([key, value]) => {
+          formData.append(key, value);
+        });
+        return formData;
+      }
+      case 'arrayBuffer':
+        return body;
+      case 'text':
+      default:
+        return body;
+    }
+  }
+}
+
+function createOfflineRequestHandler(): OfflineRequestHandler {
```

Comment:
we may not need this function, could do something like:

const offlineRequestHandler = new OfflineRequestHandler(...);